dali/pipeline/operators/reader/reader_op.h

```diff
@@ -70,49 +66,37 @@ class DataReader : public Operator<Backend> {
   }
 
   // perform the prefetching operation
-  virtual bool Prefetch() {
-    // first clear the batch
-    // TODO(slayton): exchange multiple batches
+  virtual void Prefetch() {
+    prefetched_batch_.reserve(Operator<Backend>::batch_size_);
     prefetched_batch_.clear();
 
     for (int i = 0; i < Operator<Backend>::batch_size_; ++i) {
       auto* t = loader_->ReadOne();
       prefetched_batch_.push_back(t);
     }
-
-    // TODO(slayton): swap prefetched batches around
-
-    return true;
   }
 
   // Main prefetch work loop
   void PrefetchWorker() {
     std::unique_lock<std::mutex> lock(prefetch_access_mutex_);
 
     // if a result is already ready, wait until it's consumed
-    while (prefetch_ready_) {
-      producer_.wait(lock);
-    }
+    producer_.wait(lock, [&]() { return !prefetched_batch_ready_.load(); });
 
     while (!finished_) {
       try {
-        prefetched_batch_.reserve(Operator<Backend>::batch_size_);
-        prefetch_success_ = Prefetch();
+        Prefetch();
       } catch (const std::exception& e) {
-        printf("Prefetch Failed\n");
+        std::cout << "Prefetch Failed" << std::endl;
```

Comment:
As a diagnostic output, shouldn't it go to stderr?

Reply:
We should actually pass it to DALI_FAIL

Reply:
Done